### PR TITLE
Add result aggregation for query templates

### PR DESCRIPTION
**Pull request by nck-mlcnv**
The new default behaviour is, that each result from the instances of a query template will be counted as the result of the query template. This is done by forcing the query instances to use the id of the query template in the results.
The query templates are considered as queries as well, but they won't be executed.

The ids of those queries that won't be presented in the results (instance queries if `individualResults` is turned on, query templates otherwise) will be shifted to the right, outside of the range of the number of queries whose results will be presented.

The serialization of the query instaces needed to be changed as well, so that the query instances can be mapped back to their
templates.

The query handler also needs to differentiate between the number of executable queries and the number of queries, that
will be presented in the results, otherwise, too few queries might be executed.

Todo:
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Comment by nck-mlcnv** [comment]
Also, the result model probably needs to be changed, so that it includes the relation between query templates and their instances.

File: graalvm/suite.yml [hunk 1]
```diff
@@ -6,7 +6,7 @@ connections:
   - name: "Blazegraph"
     version: "1.1.1"
     dataset: "DatasetName"
-    endpoint: "http://localhost:9999/blazegraph/sparql"
+    endpoint: "https://dbpedia.org/sparql"
```
Comment (bigerl) [inline graalvm/suite.yml]: I think it would be better to use the [wikidata endpoint](https://query.wikidata.org/). I see it as more long-living. (same below)

File: src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java [hunk 2]
```diff
@@ -141,18 +144,18 @@ public String value() {
...
                 this.save = save == null || save;
+                this.individualResults = individualResults != null && individualResults;
             }
         }
     }
 
-    public record QueryStringWrapper(int index, String query, boolean update) {}
-
-    public record QueryStreamWrapper(int index, boolean cached, Supplier<InputStream> queryInputStreamSupplier, boolean update) {}
+    public record QueryStringWrapper(int index, String query, boolean update, Integer resultId) {}
```
Comment (bigerl) [inline src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java:156]: document what the fields are for. Especially, the resultId is not self-explanatory.

File: src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java [hunk 3]
```diff
@@ -141,18 +144,18 @@ public String value() {
...
+    public record QueryStreamWrapper(int index, boolean cached, Supplier<InputStream> queryInputStreamSupplier, boolean update, Integer resultId) {}
```
Comment (bigerl) [inline src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java:170]: same here

File: src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java [hunk 4]
```diff
@@ -161,7 +164,10 @@ public record QueryStreamWrapper(int index, boolean cached, Supplier<InputStream
     final protected Config config;
 
     final protected QueryList queryList;
-    final protected List<QueryData> queryData;
+    protected List<QueryData> queryData;
+
+    int executableQueryCount = 0;     // stores the number of queries that can be executed
+    int representativeQueryCount = 0; // stores the number of queries that are represented in the results
```
Comment (bigerl) [inline src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java]:
```suggestion
    int representedQueryCount = 0; // stores the number of queries that are represented in the results
```

File: src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java [hunk 5]
```diff
@@ -191,54 +197,139 @@ public QueryHandler(Config config) throws IOException {
...
+        AtomicInteger templateIndex = new AtomicInteger(0); // index of the next template
+        AtomicInteger index = new AtomicInteger(0);      // index of the current query
+        AtomicInteger instanceId = new AtomicInteger(0); // id of the current instance for the current template
+        queryData = templateData.queries.stream().map(
+                query -> {
+                    // If "individualResults" is turned on, move the query templates outside the range of
+                    // "representativeQueryCount" to avoid them being represented in the results.
+                    // Otherwise, if "individualResults" is turned off, the instances need to be moved outside the range
+                    // of "representativeQueryCount", but because "instantiateTemplateQueries" already appends the
+                    // instances to the end of the original queries, this will already be done.
```
Comment (bigerl) [inline src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java]: Please double check and rephrase. It sounds like it must be moved `outside` either way.

File: src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java [hunk 6]
```diff
@@ -191,54 +197,139 @@ public QueryHandler(Config config) throws IOException {
...
+
+                    // once the template instances start, the template index is reset and reused for the instances
+                    // to track to which template the instances belong
```
Comment (bigerl) [inline src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java]:
```suggestion
                    // track where the template instances belong
```

File: src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java [hunk 7]
```diff
@@ -191,54 +197,139 @@ public QueryHandler(Config config) throws IOException {
...
         }
-        return (config.caching()) ?
-                new FileCachingQueryList(querySource) : // if caching is enabled, cache the instances
-                new FileReadingQueryList(querySource);  // if caching is disabled, read the instances from the file every time
+
+        // initialize queryData based on the template data
+        AtomicInteger templateIndex = new AtomicInteger(0); // index of the next template
+        AtomicInteger index = new AtomicInteger(0);      // index of the current query
+        AtomicInteger instanceId = new AtomicInteger(0); // id of the current instance for the current template
+        queryData = templateData.queries.stream().map(
```
Comment (bigerl) [inline src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java]: I do not really get what is happening in here. That needs some more documentation. Might be that updating the comment in 282 is enough to clarify.

File: src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java [hunk 8]
```diff
@@ -191,54 +197,139 @@ public QueryHandler(Config config) throws IOException {
...
+
+                    if (index.get() >= templateData.instanceStart) {
+                        // query is an instance of a template
+
+                        // if the instance id is equal to the number of instances for the current template,
+                        // the next template is used
+                        if (instanceId.get() == templateData.instanceNumber[templateIndex.get()]) {
+                            templateIndex.getAndIncrement();
+                            instanceId.set(0);
+                        }
```
Comment (bigerl) [inline src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java]: Where is instanceId ever increased? It seems to be 0 all the time. 
Also, this looks a bit like the whole if statement might require a lock for synchronization.

File: src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java [hunk 9]
```diff
@@ -191,54 +197,139 @@ public QueryHandler(Config config) throws IOException {
...
                         writer.newLine();
                     }
                 }
-                // create a new query source based on the new instance file
-                querySource = createQuerySource(instancePath);
-            } else {
-                // query source isn't necessary, because queries aren't stored in a file,
-                // directly return a list of the instances instead
-                return new StringListQueryList(instances);
             }
```
Comment (bigerl) [inline src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java]: Please extract reading and writing template headers. Best would be a separate file. 

The same applies probably to template data in general. It currently makes the QueryHandler.java file hard to follow. Would be better to refactor and extract the handling of templates into a separate file.

File: src/main/java/org/aksw/iguana/cc/query/QueryData.java [hunk 10]
```diff
@@ -12,9 +13,15 @@
...
-public record QueryData(int queryId, boolean update) {
+public record QueryData(int queryId, QueryType type, Integer templateId) {
+    public enum QueryType {
+        DEFAULT,
+        UPDATE,
+        TEMPLATE,
+        TEMPLATE_INSTANCE
+    }
+
     public static List<QueryData> generate(Collection<InputStream> queries) {
```
Comment (bigerl) [inline src/main/java/org/aksw/iguana/cc/query/QueryData.java:33]: add a docstring that explictly states that only default and update queries can be processed.

File: src/main/java/org/aksw/iguana/cc/query/QueryData.java [hunk 11]
```diff
@@ -12,9 +13,15 @@
...
  * @param queryId The id of the query
- * @param update  If the query is an update query
  */
-public record QueryData(int queryId, boolean update) {
+public record QueryData(int queryId, QueryType type, Integer templateId) {
+    public enum QueryType {
+        DEFAULT,
+        UPDATE,
+        TEMPLATE,
+        TEMPLATE_INSTANCE
```
Comment (bigerl) [inline src/main/java/org/aksw/iguana/cc/query/QueryData.java:22]: Shoudn't this be 
```suggestion
        TEMPLATE_INSTANCE_DEFAULT,
        TEMPLATE_INSTANCE_UPDATE
```

File: src/main/java/org/aksw/iguana/cc/tasks/impl/StresstestResultProcessor.java [hunk 12]
```diff
@@ -128,7 +128,8 @@ public void calculateAndSaveMetrics(Calendar start, Calendar end) {
             m.add(workerRes, RDF.type, IONT.worker);
             m.add(workerRes, IPROP.workerID, toInfinitePrecisionIntegerLiteral(worker.getWorkerID()));
             m.add(workerRes, IPROP.workerType, ResourceFactory.createTypedLiteral(worker.getClass().getSimpleName()));
-            m.add(workerRes, IPROP.noOfQueries, toInfinitePrecisionIntegerLiteral(config.queries().getQueryCount()));
+            // TODO: is it executable or representative?
```
Comment (bigerl) [inline src/main/java/org/aksw/iguana/cc/tasks/impl/StresstestResultProcessor.java]: depends on `individualResults = false/true/

File: src/main/java/org/aksw/iguana/cc/query/QueryData.java [hunk 13]
```diff
@@ -25,8 +32,24 @@ public static List<QueryData> generate(Collection<InputStream> queries) {
...
-            queryData.add(new QueryData(i++, update));
+            queryData.add(new QueryData(i++, update ? QueryType.UPDATE : QueryType.DEFAULT, null));
+            try {
+                query.close();
+            } catch (IOException ignored) {}
         }
         return queryData;
     }
+
+    public static boolean checkUpdate(InputStream query) {
```
Comment (bigerl) [inline src/main/java/org/aksw/iguana/cc/query/QueryData.java]: add a docstring

File: src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java [hunk 14]
```diff
@@ -161,7 +164,10 @@ public record QueryStreamWrapper(int index, boolean cached, Supplier<InputStream
     final protected Config config;
 
     final protected QueryList queryList;
-    final protected List<QueryData> queryData;
+    protected List<QueryData> queryData;
+
+    int executableQueryCount = 0;     // stores the number of queries that can be executed
```
Comment (bigerl) [inline src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java]: mention in the docstrings of executableQueryCount and representedQueryCount when they are the same (no template or setting explicitly  template's `individualResults: true`) and how these differ for templates.

File: src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java (hunk 4 above)
Comment (bigerl) [inline src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java]: give them a clear visibility instead of implicit package private.

**Review by bigerl — CHANGES_REQUESTED** [review]
Thx for the PR. The changes to introduce the aggregation currently seem quite complex. See my comments for suggestions on how the code could be untangled a bit.

File: src/main/java/org/aksw/iguana/cc/query/QueryData.java (hunk 11 above)
Comment (nck-mlcnv) [reply src/main/java/org/aksw/iguana/cc/query/QueryData.java:22]: No, I don't think so. `TEMPLATE` stands for the query template.

File: graalvm/suite.yml (hunk 1 above)
Comment (bigerl) [reply graalvm/suite.yml]: We've just had an issue with the wikidata endpoint because it requires the [User-Agent](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/User-Agent) to be set, see: https://foundation.wikimedia.org/wiki/Policy:User-Agent_policy

Comment (nck-mlcnv) [reply graalvm/suite.yml]: The Apache HttpClient automatically sends that header.

File: src/main/java/org/aksw/iguana/cc/query/QueryData.java (hunk 11 above)
Comment (nck-mlcnv) [reply src/main/java/org/aksw/iguana/cc/query/QueryData.java:22]: nvm

Comment (nck-mlcnv) [reply src/main/java/org/aksw/iguana/cc/query/QueryData.java:22]: actually, there is nothing that really supports update queries as templates, so i don't know if that was really necessary

File: src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java (hunk 8 above)
Comment (nck-mlcnv) [reply src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java]: Why should it be synchronized? There is no concurrency happening.

File: src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java (hunk 9 above)
Comment (nck-mlcnv) [reply src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java]: I haven't extracted the header writing part to a separate file yet.

**Comment by nck-mlcnv** [comment]
The version check test seems to be broken, I have to fix it soon. Also I haven't extracted the header writing part to a separate file yet.

File: src/main/java/org/aksw/iguana/cc/query/handler/TemplateHandler.java [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Random;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+public class TemplateHandler {
```
Comment (bigerl) [inline src/main/java/org/aksw/iguana/cc/query/handler/TemplateHandler.java]: Please add a docstring and rename to QueryTemplateManager

**Review by bigerl — CHANGES_REQUESTED** [review]
Thx for the overhaul. Code looks good now. Not much to add. Only this small documentation and renaming request below. :)
